### ISSUE TITLE
Additional clutter memory allocated on GPU 0 when training on GPU 1+

### ISSUE THREAD

**Issue by ibmua** [pytorch/pytorch#44248]
## 🐛 Bug

Might be related to https://discuss.pytorch.org/t/model-always-allocate-some-memories-on-gpu-0/58340/8

I get similar issue forking via multiprocessing and processing on seperate gpus in separate processes. Additional memory clutter is being allocated on GPU 0 (233mb in my case, seemingly independent of the model depth or breadth) during first training-mode (but not testing-mode) model forward execution.

I'm using a hack like
`
		with torch.cuda.device(1):
			recon_batch, *_ 	= model(data)
`
to get around that. No such additional persistent memory allocation takes place if I use this (not on GPU 0 nor any other GPU) or simply in case that the device in question is GPU 0. So this memory allocation is purely unnecessary clutter. Model in question is pure and minimal, composed of nn.Sequential and conv, BN, linear and inplace ReLU activations, so it's internals are most likely not the cause.

This is not related to GPU 0 specifically, but to the current default GPU. If I change gpu context to other GPU that the model is not allocated on, this memory gets allocated on this other context-GPU. 

## Environment

PyTorch version: 1.7.0.dev20200811
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 440.33.01
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.6.post0
[pip3] numpy==1.19.1
[pip3] torch==1.7.0.dev20200811
[pip3] torchvision==0.8.0.dev20200811
[conda] Could not collect

**Comment by SsnL** [comment]
You are using some operation that initializes the CUDA context without specifying a device, which initialized the context on device 0. This is not really a pytorch bug, but you could figure out which line does that.

**Comment by ag14774** [comment]
Could be pin_memory in your data loader. Does it happen if pin_memory is False?

**Comment by ibmua** [comment]
> Could be pin_memory in your data loader. Does it happen if pin_memory is False?

No, I turned that one off a while ago and transferred my dataset directly to GPU, feeding it through a custom class. As far as I remember, pin_memory was only making things load longer vs not using it, btw. 🤷



> You are using some operation that initializes the CUDA context without specifying a device, which initialized the context on device 0. This is not really a pytorch bug, but you could figure out which line does that.

Can't think of where that would happen outside of stock pytorch code. I don't have any custom code there that inits torch tensors on model->forward. And this happens during train-time model->forward.

**Comment by ngimel** [comment]
Do you have a minimum reproducing script?
